compiler/dcalc/from_scopelang.ml

```diff
@@ -81,23 +86,66 @@ let pos_mark_mk (type a m) (e : (a, m mark) gexpr) :
   let pos_mark_as e = pos_mark (Marked.get_mark e) in
   pos_mark, pos_mark_as
 
-let merge_defaults caller callee =
-  let caller =
-    let m = Marked.get_mark caller in
-    let pos = Expr.mark_pos m in
-    Expr.make_app caller
-      [Expr.elit LUnit (Expr.with_ty m (Marked.mark pos (TLit TUnit)))]
-      pos
-  in
-  let body =
-    let m = Marked.get_mark callee in
-    let ltrue =
-      Expr.elit (LBool true)
-        (Expr.with_ty m (Marked.mark (Expr.mark_pos m) (TLit TBool)))
+let merge_defaults
+    ~(is_func : bool)
+    (caller : (dcalc, 'm mark) boxed_gexpr)
+    (callee : (dcalc, 'm mark) boxed_gexpr) : (dcalc, 'm mark) boxed_gexpr =
+  (* the merging of the two defaults, from the reentrant caller and the callee,
+     is straightfoward in the general case and a little subtler when the
+     variable being defined is a function. *)
+  if is_func then
+    let m_callee = Marked.get_mark callee in
+    Bindlib.unbox
+      (Bindlib.box_apply
```

Comment:
Additionally, if you want to avoid the `Expr.rebox`, it might be possible to keep the `Bindlib.box_apply` (or `Expr.Box.app1` !) but unbox the result of the inner `Expr.make_abs` instead. That's assuming the free variables remain untouched by the transformation of course.

Reply:
Thanks for the advice!